### PR TITLE
mgmt, support webapp flex one deploy

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
fix https://github.com/Azure/azure-sdk-for-java/issues/40987

test is record/live only, tested and passes

# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/models/FunctionsScaleAndConcurrency.java
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/models/FunctionsScaleAndConcurrency.java:1]: This is a temporary change, pending for spec fix in https://github.com/Azure/azure-rest-api-specs/pull/29907

The public type of this 2 properties should be Integer.

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/models/FunctionApp.java [hunk 1]
```diff
@@ -146,6 +147,21 @@ public interface FunctionApp extends FunctionAppBasic, WebAppBase, Updatable<Fun
...
+     * @return the completion of the operation
+     */
+    Mono<Void> deployAsync(File file);
+
+    /**
+     * Deploy a zip file to Azure site, for Flex Consumption plan.
+     *
+     * @param file the file to upload
+     */
+    void deploy(File file);
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/models/FunctionApp.java:163]: Add a minimal support, as Flex Consumption plan feature is still preview.

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/test/java/com/azure/resourcemanager/appservice/OneDeployTests.java [hunk 2]
```diff
@@ -116,6 +136,82 @@ public void canPushDeployJar() throws Exception {
         waitForRuntimeSuccess(slot2, slotDeploymentId);
     }
 
+    // test uses storage account key and connection string to configure the function app
+    @Test
+    @DoNotRecord(skipInPlayback = true)
+    public void canDeployFlexConsumptionFunctionApp() {
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-appservice/src/test/java/com/azure/resourcemanager/appservice/OneDeployTests.java:142]: nit, should we put this test into `FunctionAppTests`? As far as I can see, FunctionApp doesn't implement `SupportsOneDeploy`.

**Review by XiaofeiCao — APPROVED** [review]

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/test/java/com/azure/resourcemanager/appservice/OneDeployTests.java (hunk 2 above)
Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/test/java/com/azure/resourcemanager/appservice/OneDeployTests.java:142]: The endpoint `api/publish` is the OneDeploy endpoint.

I guess my mistake was the naming of `SupportsOneDeploy` interface. From what we see currently, it is only `SupportsWebAppOneDeploy`...

Let me see if I can get some clarification from service, about what is their terminology on `api/publish` endpoint.

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java [hunk 3]
```diff
@@ -807,6 +816,137 @@ Mono<ConnectionStringDictionaryInner> updateConnectionStrings(ConnectionStringDi
...
+                    return kuduClient.pushZipDeployAsync(file)
+                        // it appears that "getDeploymentStatus" must first be called with deploymentId="latest"
+                        // this "latest" also works, if polling is done via kudu "api/deployments/{deploymentId}" API
+                        .then(Mono.just(new KuduDeploymentResult("latest")));
+                }
+            } catch (IOException e) {
+                return Mono.error(e);
+            }
+        });
+    }
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java:888]: `pushDeploy` branches on tier.

For non-Flex, it calls zipDeploy
For Flex, it calls oneDeploy

There is actually another branching for ConsumptionPlan (non-Flex) that CLI uses "WEBSITE_RUN_FROM_PACKAGE"

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/KuduClient.java [hunk 4]
```diff
@@ -324,19 +360,56 @@ Mono<KuduDeploymentResult> pushDeployAsync(DeployType type, File file, String pa
...
+        return getDeploymentResult(retryOnError(service.zipDeploy(host, flux, length,
+            true, DEPLOYER_JAVA_SDK)), true);
+    }
+
+    Mono<KuduDeploymentResult> pushDeployFlexConsumptionAsync(File file) throws IOException {
+        AsynchronousFileChannel fileChannel = AsynchronousFileChannel.open(file.toPath(), StandardOpenOption.READ);
+        return retryOnError(service.deployFlexConsumption(host, FluxUtil.readFile(fileChannel), fileChannel.size(),
+            false, DEPLOYER_JAVA_SDK))
+            // there is no "SCM-DEPLOYMENT-ID" header in response
+            .then(Mono.just(new KuduDeploymentResult("latest")))
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/KuduClient.java:412]: Why we only retry on flex case, not normal zip case?

**Review by XiaofeiCao — APPROVED** [review]

**Pull request merged**